### DISCUSSION TITLE
How to let Play/Guice print the unsatisfied dependencies while bootstrap?

### DISCUSSION THREAD

**Discussion by magicliang** [playframework/playframework#11908] [Q&A]
Hi, guys.

My question is: is there any property or config I can use to command Play application to tell me what dependency is missed while bootstrap?

I have a complex legacy Play-based web application, which merged a lot of commits recently, and can not be bootstrapped now. Per my understanding, there must be some unsatisfied "@Inject" was introduced, but the commits are too many, I can't figure out which commit to blame.

We use Guice in several flavors:
1. We use @Singleton annotation.
2. We use bind api in a big Custom Guice module like:
```java
public class GuiceModule extends AbstractModule {

   @Override
    protected void configure() {
         
        bindConstant().annotatedWith(Names.named("DIMaxInQuery"))
            .to(config.getInt("di.options.maxInQuery"));
   }
}
```

and we inject dependencies like:

```java
    public cloudPulsarClient(@Named("cloudPulsarServiceUrl") String serviceUrl,
            @Named("primaryConfig") Config config) throws PulsarClientException {
        this.serviceUrl = serviceUrl;
        this.authorizationToken = config.getString("cloud.pulsar.secret");
        this.pulsarClient = initClient();
    }
```

The application just fails silently, and no info printed to show what is missed.

We hope some config to print the info, like what is done in Spring framework:

```
    org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'customerController' defined in file [D:\Coding\Java\Spring_Framework\zzz\target\zzz\WEB-INF\classes\com\bai\controller\CustomerController.class]: Unsatisfied dependency expressed through constructor parameter 0; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'customerServiceImpl': Unsatisfied dependency expressed through field 'customerDao'; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'customerDaoImpl' defined in file [D:\Coding\Java\Spring_Framework\zzz\target\zzz\WEB-INF\classes\com\bai\dao\CustomerDaoImpl.class]: Unsatisfied dependency expressed through constructor parameter 0; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'getSessionFactory' defined in com.bai.config.HibernateConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.hibernate.SessionFactory]: Factory method 'getSessionFactory' threw exception; nested exception is java.lang.NoClassDefFoundError: javax/xml/bind/JAXBException



```

**Comment by Nezisi** [comment]
nested exception is java.lang.NoClassDefFoundError: javax/xml/bind/JAXBException

Java EE and CORBA modules are deprecated in JDK 9 and 10, and are removed completely in JDK 11 and later.

Thus you need to add manually a JAXB implementation to the runtime, though I'd check with a magnifying glass the dependencies.

You didn't mention any kind of version afaik, but I'd highly recommend you look for updates.

Most libraries have released newer upstream versions to migrate to Jakarta. These updated versions should come in large quantity with Play Framework 2.9 - but it should be safe to update them in Play 2.8.

I'd recommend upgrading, especially to prepare for Play 2.9, as the Jakarta migration is required for getting to JDK 17 plus.

JDK 21 as the next LTS release is around the corner ;)

Hope that helps, good luck!

**Reply by magicliang** [reply]
Maybe I set a wrong example here.

My point is, sometimes we didn't create necessary dependency, but the framework stopped silently without printing anything like what Spring does. What I tried to show in that example is, Spring will print what exact dependency is not satisfied, so we can fix the problem. 

In my case, the missed dependency is like this:
```java
public class CustomProducer {


}

public class CustomerService {

@Inject
private CustomProducer producer;
}
```

My colleague forgot to add @Singleton annotation to CustomProducer. So Play application stopped. In Spring, some dependency exception will be thrown, that's far more friendly to developers.

We fixed our issue finally like this:

```java
@Singleton
public class CustomProducer {


}
```

**Reply by Nezisi** [reply]
My bad.

I'm not sure if Guide detects *missing* Inject annotations, but there were improvements in Guice 6 regarding error output.

As far as I know, Play has no mechanism for that.

**Reply by mkurz** [reply]
I can not see why "forgetting" adding `@Singleton` to a object that will be injected makes the framework fail. Doesn't make sense IMHO. If the object is not a singleton, than new instances will be created each time injected, so this shouldn't really matter.
Maybe there is another problem, not related to Play or Guice?
